gapis/replay/executor/executor.go

```diff
@@ -113,105 +113,111 @@ func (r executor) execute(ctx context.Context) error {
+			if err := e.handleGetData(ctx, r, w); err != nil {
 				return fmt.Errorf("Failed to read replay postback data: %v", err)
 			}
 		case protocol.MessageType_Post:
-			if err := r.handleDataResponse(ctx, postbacks); err != nil {
+			if err := e.handleDataResponse(ctx, r, postbacks); err != nil {
 				return fmt.Errorf("Failed to send replay resource data: %v", err)
 			}
 		default:
-			return fmt.Errorf("Unknown message type: %v\n", msg)
+			return fmt.Errorf("Unknown message type: %v", msg)
+		}
+
+		if err := bw.Flush(); err != nil {
+			return err
 		}
 	}
 }
 
-func (r executor) handleDataResponse(ctx context.Context, postbacks io.Writer) error {
-	d := endian.Reader(r.connection, r.memoryLayout.GetEndian())
-
-	n := d.Uint32()
-	if d.Error() != nil {
-		return d.Error()
+func (e executor) handleDataResponse(ctx context.Context, r binary.Reader, postbacks io.Writer) error {
+	n := r.Uint32()
+	if r.Error() != nil {
+		return r.Error()
 	}
 
-	c, err := io.CopyN(postbacks, r.connection, int64(n))
+	c, err := io.CopyN(postbacks, r, int64(n))
 	if c != int64(n) {
 		return err
 	}
 
 	return nil
 }
 
-func (r executor) handleGetData(ctx context.Context) error {
+func (e executor) handleGetData(ctx context.Context, r binary.Reader, w binary.Writer) error {
 	ctx = log.Enter(ctx, "handleGetData")
-	d := endian.Reader(r.connection, r.memoryLayout.GetEndian())
 
-	resourceCount := d.Uint32()
-	if err := d.Error(); err != nil {
+	resourceCount := r.Uint32()
+	if err := r.Error(); err != nil {
 		return log.Err(ctx, err, "Failed to decode resource count")
 	}
 
-	totalExpectedSize := d.Uint64()
-	if err := d.Error(); err != nil {
+	totalExpectedSize := r.Uint64()
+	if err := r.Error(); err != nil {
 		return log.Err(ctx, err, "Failed to decode total expected size")
 	}
 
-	resourceIDs := make([]id.ID, resourceCount)
-	for i := range resourceIDs {
-		idString := d.String()
-		if d.Error() != nil {
-			return d.Error()
-		}
-		var err error
-		resourceIDs[i], err = id.Parse(idString)
-		if err != nil {
-			return log.Errf(ctx, err, "Failed to parse resource id: %v", idString)
+	totalReturnedSize := 0
+
+	buf := bytes.Buffer{}
+	db := database.Get(ctx)
+	for i := uint32(0); i < resourceCount; i++ {
+		idString := r.String()
+		if r.Error() != nil {
+			return r.Error()
 		}
-	}
 
-	totalReturnedSize := uint64(0)
-	for _, rid := range resourceIDs {
-		obj, err := database.Resolve(ctx, rid)
+		rID, err := id.Parse(idString)
 		if err != nil {
-			return log.Errf(ctx, err, "Failed to resolve resource with id: %v", rid)
+			return log.Errf(ctx, err, "Failed to parse resource id: %v", idString)
 		}
 
-		data := obj.([]byte)
-		n, err := r.connection.Write(data)
+		obj, err := db.Resolve(ctx, rID)
 		if err != nil {
-			return log.Errf(ctx, err, "Failed to send resource with id: %v", rid)
+			return log.Errf(ctx, err, "Failed to resolve resource with id: %v", rID)
 		}
 
-		totalReturnedSize += uint64(n)
+		n, _ := buf.Write(obj.([]byte))
```

Comment:
This was intentional as I wanted to reduce the number of syscalls on the sending side to a minimum. I did however add the `bufio.NewWriter(connection)` line after this, which should reduce the calls significantly (still more than once by just building a single buffer).
Done anyway.